### PR TITLE
Support for abuseIPDB

### PR DESCRIPTION
**Pull request by zbalkan**
Currently failing some tests unrelated to abuseIPDB -it seems. When everything is fixed, I will update the documentation too.

### CODE REVIEW

**Comment by pirxthepilot** [comment]
@zbalkan thanks for the PR. Pretty cool! Looking forward to your changes. I'm a bit swamped this week but will review as soon as I can.

**Comment by pirxthepilot** [comment]
Btw, to run the tests locally you will need the `hatch` package. Then you can run them all:

```
hatch run test-all
```

You can also run each test separately:

Lint:

```
hatch run lint:all
```

Typechecking:

```
hatch run typecheck
```

Actual tests:

```
hatch run test
```

```
hatch run test-cov
```

**Comment by zbalkan** [comment]
I will solve the test problems first. Then I must write new ones.

```js
FAILED tests/test_cli.py::TestDefaults::test_defaults_3 - SystemExit: 2
FAILED tests/test_cli.py::TestDefaults::test_defaults_4 - SystemExit: 2
FAILED tests/test_cli.py::TestGenEntityHandler::test_handler_domain_1 - assert False
FAILED tests/test_cli.py::TestGenEntityHandler::test_handler_domain_2 - assert False
FAILED tests/test_cli.py::TestGenEntityHandler::test_handler_domain_3 - assert False
FAILED tests/test_cli.py::TestGenEntityHandler::test_handler_domain_4 - assert False
FAILED tests/test_cli.py::TestGenEntityHandler::test_handler_ip_1 - assert False
FAILED tests/test_cli.py::TestMain::test_main_default - AssertionError: expected call not found.
============================================================= 8 failed, 105 passed in 4.53s ============================================================== 
```




When it comes to the UI, there will only be a single line just like Greynoise has.
```python
        #
        # Content
        #
        benign_style = self.theme.info
        malicious_style = self.theme.warn

        text = Text()

        score = f"{str(ip.abuseConfidenceScore)} abuse confidence score"
        text.append(
            Text(score, style=malicious_style)
            if ip.abuseConfidenceScore > 0
            else
            Text(score, style=benign_style))
```

So the expected result is green or red depending on the score.

**Review by pirxthepilot — CHANGES_REQUESTED** [review]
@zbalkan just finished my initial review. Apologies for the multiple syntax/formatting changes. I didn't think anyone would actually contribute code here so I never got around to autoformatting and style guides yet 😅. But thanks for taking the time and putting up with my messy code!

File: tests/test_cli.py [hunk 1]
```diff
@@ -23,24 +18,19 @@
...
 from wtfis.handlers.ip import IpAddressHandler
-from wtfis.main import (
-    generate_entity_handler,
-    generate_view,
-    main,
-    parse_args,
-    parse_env,
-)
+from wtfis.main import (generate_entity_handler, generate_view, main,
+                        parse_args, parse_env)
```
[the remark this answers is not in the document]
Comment (zbalkan) [reply tests/test_cli.py:22]: I used [isort](https://marketplace.visualstudio.com/items?itemName=ms-python.isort) for import formatting. It's my go-to for a while. for other formatting I used autopep8 but it is not a big deal I believe. It's your preference in the end

File: tests/test_cli.py [hunk 2]
```diff
@@ -380,7 +372,7 @@ def test_handler_domain_1(self, fake_load_dotenv_1):
             parse_env()
             console = Console()
             progress = simulate_progress(console),
-            entity = generate_entity_handler(parse_args(), console, progress)
+            entity = generate_entity_handler(parse_args(), console, progress)  # type: ignore
```
[the remark this answers is not in the document]
Comment (zbalkan) [reply tests/test_cli.py]: Done.

File: tests/test_cli.py [hunk 3]
```diff
@@ -493,7 +487,7 @@ def test_view_ip_1(self, m_ip_view, test_data):
     def test_view_error(self):
         """ IP address view with default params """
         with pytest.raises(Exception):
-            generate_view(MagicMock(), MagicMock(), "foobar")
+            generate_view(MagicMock(), MagicMock(), "foobar")  # type: ignore
```
[the remark this answers is not in the document]
Comment (zbalkan) [reply tests/test_cli.py]: Done.

File: wtfis/clients/abuseipdb.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    @property
+    def name(self) -> str:
+        return "abuseIPDB"
+
+    def _get_ip(self, ip: str) -> Optional[abuseIPDBIp]:
+        # Let a 404 or invalid IP pass
+        try:
+            params = {"ipAddress": ip, 'maxAgeInDays': '90'}
+            headers = {"key": self.api_key, 'Accept': 'application/json'}
```
[the remark this answers is not in the document]
Comment (zbalkan) [reply wtfis/clients/abuseipdb.py]: Done.

File: wtfis/main.py [hunk 5]
```diff
@@ -129,6 +135,19 @@ def generate_entity_handler(
...
+        abuseIPDBClient(os.environ["ABUSEIPDB_API_KEY"])
+        if args.use_abuseipdb else None
+    )
+
+    # URLhaus client (optional)
+    urlhaus_client = (
+        UrlHausClient()
+        if args.use_urlhaus
+        else None
+    )
```
[the remark this answers is not in the document]
Comment (zbalkan) [reply wtfis/main.py]: Done.

File: wtfis/models/abuseipdb.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    countryName: Optional[str] = None
+    usageType: Optional[str] = None
+    isp: str
+    domain: Optional[str] = None
+    hostnames: Optional[list[str]] = None
+    isTor: Optional[bool] = None
+    totalReports: int
+    numDistinctUsers: Optional[int] = None
+    lastReportedAt: Optional[str] = None
+    reports: Optional[list] = None
```
[the remark this answers is not in the document]
Comment (zbalkan) [reply wtfis/models/abuseipdb.py]: Done.

File: wtfis/ui/base.py [hunk 7]
```diff
@@ -251,6 +254,31 @@ def _gen_greynoise_tuple(self, ip: GreynoiseIp) -> Tuple[Text, Text]:
 
         return title, text
 
+    def _gen_abuseipdb_tuple(self, ip: abuseIPDBIp) -> Tuple[Text, Text]:
+        #
+        # Title
+        #
+        title = self._gen_linked_field_name("abuseIPDB", hyperlink=f"https://www.abuseipdb.com/check/{ip.ipAddress}")
```
[the remark this answers is not in the document]
Comment (zbalkan) [reply wtfis/ui/base.py]: Done.

File: wtfis/models/abuseipdb.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+
+class abuseIPDBIpMap(RootModel):
```
[the remark this answers is not in the document]
Comment (zbalkan) [reply wtfis/models/abuseipdb.py]: Done.

**Comment by zbalkan** [comment]
I did the changes requested. Yet, I need to fix the tests. I may need some guidance there.

**Comment by zbalkan** [comment]
Since there are no "Discussions" section, I'd like to ask here. I see that there are two types of data collected, the metadata on IP or domain -aka whois- and reputation.

| Service | WHOIS | Reputation | Notes |
| --- | --- | --- | --- |
| Virustotal | + | + | |
| Passivetotal | + | - | Has API for reputation |
| IP2Whois | + | - | |
| IPWhois | + | - | |
| Shodan | + | - | |
| Greynoise | - | + | |
| URLhaus | - | + | |

Wouldn't it be nice to categorize them based on those? Though PT reputation endpoint may put it in both categories just like VT. Currently, the term "enrichment" is used yet it does not add much semantically.

**Comment by pirxthepilot** [comment]
@zbalkan sorry for the late reply, weekdays are usually busy. I will try to review tomorrow and help with the tests. Thank you for the changes!

**Comment by pirxthepilot** [comment]
> Since there are no "Discussions" section, I'd like to ask here. I see that there are two types of data collected, the metadata on IP or domain -aka whois- and reputation.
> Service 	WHOIS 	Reputation 	Notes
> Virustotal 	+ 	+ 	
> Passivetotal 	+ 	- 	Has API for reputation
> IP2Whois 	+ 	- 	
> IPWhois 	+ 	- 	
> Shodan 	+ 	- 	
> Greynoise 	- 	+ 	
> URLhaus 	- 	+ 	
> 
> Wouldn't it be nice to categorize them based on those? Though PT reputation endpoint may put it in both categories just like VT. Currently, the term "enrichment" is used yet it does not add much semantically.

Good call out. Technically, these are all enrichments anyway. I like the categorization, but I wonder if there's a better term to describe the non-reputation data other than "whois" (though, I'm not opposed to the term).

File: wtfis/clients/abuseipdb.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
+from typing import Optional
+
+from requests.exceptions import HTTPError
+
+from wtfis.clients.base import BaseIpEnricherClient, BaseRequestsClient
+from wtfis.models.abuseipdb import AbuseIpDb, AbuseIpDbMap
+
+
+class abuseIPDBClient(BaseRequestsClient, BaseIpEnricherClient):
```
Comment (pirxthepilot) [inline wtfis/clients/abuseipdb.py:9]: Oops, I missed this one. Can you change to CamelCase, i.e. `AbuseIpDbClient` ?

**Comment by zbalkan** [comment]
> > Since there are no "Discussions" section, I'd like to ask here. I see that there are two types of data collected, the metadata on IP or domain -aka whois- and reputation.
> > Service 	WHOIS 	Reputation 	Notes
> > Virustotal 	+ 	+
> > Passivetotal 	+ 	- 	Has API for reputation
> > IP2Whois 	+ 	-
> > IPWhois 	+ 	-
> > Shodan 	+ 	-
> > Greynoise 	- 	+
> > URLhaus 	- 	+
> > Wouldn't it be nice to categorize them based on those? Though PT reputation endpoint may put it in both categories just like VT. Currently, the term "enrichment" is used yet it does not add much semantically.
> 
> Good call out. Technically, these are all enrichments anyway. I like the categorization, but I wonder if there's a better term to describe the non-reputation data other than "whois" (though, I'm not opposed to the term).

Well, I agree with "whois" not being a good term. I just wanted to put things in a frame. You can call it any term you want. I thought of "basics" and "metadata" but didn't like it. So I kept it that way.

I agree that they are all enrichments,  the ownership is a matter of registration and static while reputation is a collective measure from multiple reports and security products and dynamic. The nature of the data is different. I wanted to point out the nuance.

**Review by pirxthepilot — CHANGES_REQUESTED** [review]

File: wtfis/clients/abuseipdb.py (hunk 9 above)
Comment (zbalkan) [reply wtfis/clients/abuseipdb.py]: Done.

File: wtfis/models/abuseipdb.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    is_public: Optional[bool] = Field(None, alias="isPublic")
+    ip_version: Optional[int] = Field(None, alias="numDistinctUsers")
+    is_whitelisted: Optional[bool] = Field(None, alias="ipVersion")
+    abuse_confidence_score: int = Field(alias="abuseConfidenceScore")
+    country_code: Optional[str] = Field(None, alias="countryCode")
+    country_name: Optional[str] = Field(None, alias="countryName")
+    usage_type: Optional[str] = Field(None, alias="usageType")
+    isp: str
+    domain: Optional[str] = None
+    hostnames: Optional[list[str]] = None
```
Comment (pirxthepilot) [inline wtfis/models/abuseipdb.py:19]: Should be `List[str]` and not `list[str]`. I think this is what's causing the "type object is not subscriptable" error in the tests.

File: wtfis/models/abuseipdb.py [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    is_tor: Optional[bool] = Field(None, alias="isTor")
+    total_reports: Optional[int] = Field(None, alias="totalReports")
+    num_distinct_users: Optional[int] = Field(None, alias="numDistinctUsers")
+    last_reported_at: Optional[str] = Field(None, alias="lastReportedAt")
+    reports: Optional[list] = None
```
Comment (pirxthepilot) [inline wtfis/models/abuseipdb.py:24]: Should this also be a list of strings (`List[str]`) like in line 19?

File: wtfis/clients/abuseipdb.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    """
+    baseurl = "https://api.abuseipdb.com/api/v2/check"
+
+    def __init__(self, api_key: str) -> None:
+        super().__init__()
+        self.api_key = api_key
+
+    @property
+    def name(self) -> str:
+        return "abuseIPDB"
```
Comment (pirxthepilot) [inline wtfis/clients/abuseipdb.py:21]: Please make this `AbuseIPDB`. Thanks!

**Review by pirxthepilot — CHANGES_REQUESTED** [review]

File: wtfis/clients/abuseipdb.py (hunk 12 above)
Comment (zbalkan) [reply wtfis/clients/abuseipdb.py]: Solved.

File: wtfis/models/abuseipdb.py [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    hostnames: Optional[List[str]] = None
+    is_tor: Optional[bool] = Field(None, alias="isTor")
+    total_reports: Optional[int] = Field(None, alias="totalReports")
+    num_distinct_users: Optional[int] = Field(None, alias="numDistinctUsers")
+    last_reported_at: Optional[str] = Field(None, alias="lastReportedAt")
+    reports: Optional[List] = None
+
+
+class AbuseIpDbMap(RootModel):
+    root: dict[str, AbuseIpDb]
```
Comment (pirxthepilot) [inline wtfis/models/abuseipdb.py:28]: Sorry, I missed this one. This should be `Dict[str, AbuseIpDb]`. I'm surprised the tests are passing except in Python 3.8!

File: wtfis/models/abuseipdb.py [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    country_name: Optional[str] = Field(None, alias="countryName")
+    usage_type: Optional[str] = Field(None, alias="usageType")
+    isp: str
+    domain: Optional[str] = None
+    hostnames: Optional[List[str]] = None
+    is_tor: Optional[bool] = Field(None, alias="isTor")
+    total_reports: Optional[int] = Field(None, alias="totalReports")
+    num_distinct_users: Optional[int] = Field(None, alias="numDistinctUsers")
+    last_reported_at: Optional[str] = Field(None, alias="lastReportedAt")
+    reports: Optional[List] = None
```
Comment (pirxthepilot) [inline wtfis/models/abuseipdb.py:24]: Just checked the [docs](https://docs.abuseipdb.com/#check-endpoint) and it looks like this is a list of dicts:

```
      "reports": [
        {
          "reportedAt": "2018-12-20T20:55:14+00:00",
          "comment": "Dec 20 20:55:14 srv206 sshd[13937]: Invalid user oracle from 118.25.6.39",
          "categories": [
            18,
            22
          ],
          "reporterId": 1,
          "reporterCountryCode": "US",
          "reporterCountryName": "United States"
        }
      ]
```
Normally, I would create a new model to represent the dict, but if you're not using this field at all then you can just remove this line.

**Review by pirxthepilot — CHANGES_REQUESTED** [review]

File: wtfis/models/abuseipdb.py (hunk 13 above)
Comment (zbalkan) [reply wtfis/models/abuseipdb.py:28]: It made me think if I used vermin, would it show the problem. Sure.

File: wtfis/models/abuseipdb.py (hunk 14 above)
Comment (zbalkan) [reply wtfis/models/abuseipdb.py:24]: Well, I intended to use it but the report detail will not be used in a calculation or displayed at all. I will just remove it. For the sake of simplification, I can just remove other unused fields too then. That'd simplify the model.

File: wtfis/models/abuseipdb.py (hunk 13 above)
Comment (zbalkan) [reply wtfis/models/abuseipdb.py:28]: It was a 3.9 thing: https://docs.python.org/3/whatsnew/3.9.html#type-hinting-generics-in-standard-collections

Comment (zbalkan) [reply wtfis/models/abuseipdb.py]: Solved. I also checked against vermin:

```powershell
vermin --target=3.8 --violations --backport argparse --backport typing --backport ipaddress --backport typing_extensions  --eval-annotations --no-parse-comments .\
```

File: wtfis/models/abuseipdb.py (hunk 14 above)
Comment (zbalkan) [reply wtfis/models/abuseipdb.py]: Removed reports but kept others which may be valuable. But YAGNI tells me I should remove others. Not sure.

File: wtfis/models/abuseipdb.py (hunk 13 above)
Comment (pirxthepilot) [reply wtfis/models/abuseipdb.py]: Good catch! Also just learned about vermin - pretty cool!

File: wtfis/models/abuseipdb.py (hunk 14 above)
Comment (pirxthepilot) [reply wtfis/models/abuseipdb.py]: All good, we can keep the rest as is.

**Comment by pirxthepilot** [comment]
@zbalkan I will go ahead and merge this. Appreciate all the work here, thank you! This will be included in the next release - just give me another week or so to write additional tests and include some tweaks not related to this.

**Comment by pirxthepilot** [comment]
@zbalkan sorry again for the late response. Yes, would appreciate the update to the README! I haven't gotten around to writing the additional tests yet, unfortunately. If you do have extra time, do you want to give it a shot? I can give you some instructions on what to add. Thanks!

**Comment by zbalkan** [comment]
Just tell me what I can do. I'll try to find some time.

**Pull request merged**